drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/project_unit_tracker.rb

```diff
@@ -0,0 +1,69 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# track assignments of units to enrollments / households
+module HmisExternalApis::AcHmis::Importers::Loaders
+  class ProjectUnitTracker
+    attr_reader :assignments, :data_source
+
+    def initialize(data_source)
+      @data_source = data_source
+      projects = Hmis::Hud::Project.where(data_source: data_source)
+
+      # {enrollment_pk => [project_pk, household_id]}
+      @enrollment_lookup = {}
+      # {enrollment_pk => hoh_entry_date}
+      @enrollment_entry_dates = {}
+      enrollment_scope.preload(:project).find_each do |enrollment|
+        @enrollment_lookup[enrollment.id] = [enrollment.project.id, enrollment.household_id]
+        @enrollment_entry_dates[enrollment.id] ||= enrollment.entry_date if enrollment.head_of_household?
+      end
+
+      @household_assignments = {}
+      @assignments = {}
+
+      # we don't check if the unit is occupied, assumption is that all unit occupancies
+      # are deleted before import
+      # { [project_id, mper_id] => [unit_ids...] }
+      @unit_lookup = Hmis::Unit
+        .where(project_id: projects.select(:id))
+        .preload(unit_type: :mper_id)
+        .to_a
+        .group_by { |u| [u.project_id, u.unit_type.mper_id.value] }
+        .transform_values { |v| v.map(&:id) }
+    end
+
+    def assign_next_unit(enrollment_pk:, unit_type_mper_id:, fallback_start_date: nil)
+      raise 'missing enrollment' if enrollment_pk.nil?
+      return if assignments[enrollment_pk] || unit_type_mper_id.blank?
+
+      unit_pk = unit_pk_for_enrollment_pk(enrollment_pk, unit_type_mper_id)
+      return unless unit_pk
```

Comment:
So this would happen if we ran out of units to assign, right? Can we log a warning here that includes the enrollment pk and unit type mper id?

Reply:
I'm curious to see if this comes up in the staging test migration.

Reply:
oh yes, good point, I'll add some logging for this case. Thanks!